Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/StartupHandler.java

```diff
@@ -88,30 +90,41 @@ private static String[] getGamesFromIntent(Intent intent)
     return null;
   }
 
+  private static Instant getSessionTimestamp(Context context)
+  {
+    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
+    long timestamp = preferences.getLong(SESSION_TIMESTAMP, 0);
+    return Instant.ofEpochMilli(timestamp);
+  }
+
   /**
-   * There isn't a good way to determine a new session. setSessionTime is called if the main
-   * activity goes into the background.
+   * Called on activity stop / to set timestamp to "now".
    */
-  public static void setSessionTime(Context context)
+  public static void updateSessionTimestamp(Activity activity)
   {
+    Context context = activity.getApplicationContext();
     SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
     SharedPreferences.Editor sPrefsEditor = preferences.edit();
-    sPrefsEditor.putLong(LAST_CLOSED, System.currentTimeMillis());
+    sPrefsEditor.putLong(SESSION_TIMESTAMP, Instant.now().toEpochMilli());
     sPrefsEditor.apply();
   }
 
   /**
-   * Called to determine if we treat this activity start as a new session.
+   * Called on activity start. Generates analytics start event if it's a fresh start of the app, or
+   * if it's a start after a long period of the app not being used (during which time the process
+   * may be restarted for power/memory saving reasons, although app state persists).
    */
-  public static void checkSessionReset(Context context)
+  public static void reportStartToAnalytics(Activity activity, boolean firstStart)
   {
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    long lastOpen = preferences.getLong(LAST_CLOSED, 0);
-    final Instant current = Instant.now();
-    final Instant lastOpened = Instant.ofEpochMilli(lastOpen);
-    if (current.isAfter(lastOpened.plus(6, ChronoUnit.HOURS)))
+    Context context = activity.getApplicationContext();
+    final Instant sessionTimestamp = getSessionTimestamp(context);
+    final Instant now = Instant.now();
+    if (firstStart || now.isAfter(sessionTimestamp.plus(6, ChronoUnit.HOURS)))
     {
-      new AfterDirectoryInitializationRunner().runWithoutLifecycle(
+      // Just in case: ensure start event won't be accidentally sent too often.
+      updateSessionTimestamp(activity);
+
+      new AfterDirectoryInitializationRunner().runWithLifecycle((LifecycleOwner) activity,
```

Comment:
I would say that runWithoutLifecycle makes more sense here. Even if the user navigates away from the main activity quickly, we still want to send the analytics event. (Though, I guess in practice there's very little you can navigate to before directory initialization finishes.)